pkg/clusteragent/admission/mutate/auto_instrumentation.go

```diff
@@ -0,0 +1,180 @@
+)
+
+var (
+	tracerVersionAnnotationKeyFormat = "admission.datadoghq.com/%s-tracer.version"
+	customTracerAnnotationKeyFormat  = "admission.datadoghq.com/%s-tracer.custom-image"
+	supportedLanguages               = []string{
+		"java",
+		"node",
+	}
+)
+
+// InjectAutoInstrumentation injects APM libraries into pods
+func InjectAutoInstrumentation(rawPod []byte, ns string, dc dynamic.Interface) ([]byte, error) {
+	return mutate(rawPod, ns, injectAutoInstrumentation, dc)
+}
+
+func injectAutoInstrumentation(pod *corev1.Pod, _ string, _ dynamic.Interface) error {
+	if pod == nil {
+		return errors.New("cannot inject lib into nil pod")
+	}
+
+	if containsInitContainer(pod, initContainerName) {
+		// The admission can be reinvocated for the same pod
+		// Fast return if we injected the library already
+		log.Debugf("Init container %q already exists in pod %q", initContainerName, podString(pod))
+		return nil
+	}
+
+	language, image, shouldInject := extractLibInfo(pod, config.Datadog.GetString("admission_controller.auto_instrumentation.container_registry"))
+	if !shouldInject {
+		return nil
+	}
+
+	return injectAutoInstruConfig(pod, language, image)
+}
+
+// extractLibInfo returns the language, the image,
+// and a boolean indicating whether the library should be injected into the pod
+func extractLibInfo(pod *corev1.Pod, containerRegistry string) (string, string, bool) {
+	podAnnotations := pod.GetAnnotations()
+	for _, lang := range supportedLanguages {
+		if image, found := podAnnotations[fmt.Sprintf(customTracerAnnotationKeyFormat, lang)]; found {
+			return lang, image, true
+		}
+
+		if version, found := podAnnotations[fmt.Sprintf(tracerVersionAnnotationKeyFormat, lang)]; found {
+			image := fmt.Sprintf("%s/dd-%s-agent-init:%s", containerRegistry, lang, version)
+			return lang, image, true
+		}
+	}
+
+	return "", "", false
+}
+
+func injectAutoInstruConfig(pod *corev1.Pod, language, image string) error {
+	injected := false
+	defer func() {
+		metrics.LibInjectionAttempts.Inc(language, strconv.FormatBool(injected))
+	}()
+
+	switch strings.ToLower(language) {
+	case "java":
+		injectLibInitContainer(pod, image)
+		err := injectLibConfig(pod, javaToolOptionsKey, javaToolOptionsValue)
+		if err != nil {
+			metrics.LibInjectionErrors.Inc(language)
+			return err
+		}
+
+	case "node":
+		injectLibInitContainer(pod, image)
+		err := injectLibConfig(pod, nodeOptionsKey, nodeOptionsValue)
+		if err != nil {
+			metrics.LibInjectionErrors.Inc(language)
+			return err
+		}
+	default:
+		metrics.LibInjectionErrors.Inc(language)
+		return fmt.Errorf("language %q is not supported", language)
+	}
+
+	injectLibVolume(pod)
+	injected = true
+
+	return nil
+}
+
+func injectLibInitContainer(pod *corev1.Pod, image string) {
+	log.Debugf("Injecting init container named %q with image %q into pod %s", initContainerName, image, podString(pod))
+	pod.Spec.InitContainers = append(pod.Spec.InitContainers,
```

Comment:
as we discuss, maybe it is better to inject the init-containers at the start of the container list. to be able to use it also in other `initContainers`